src/SDK/Common/Util/ShutdownHandler.php

```diff
@@ -0,0 +1,130 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OpenTelemetry\SDK\Common\Util;
+
+use function array_key_last;
+use ArrayAccess;
+use function assert;
+use function class_exists;
+use Closure;
+use function get_class;
+use const PHP_VERSION_ID;
+use ReflectionFunction;
+use function register_shutdown_function;
+use function spl_object_id;
+use stdClass;
+use WeakReference;
+
+final class ShutdownHandler
+{
+    /** @var array<int, Closure>|null */
+    private static ?array $handlers = null;
+    /** @var ArrayAccess<object, self>|null  */
+    private static ?ArrayAccess $weakMap = null;
+    /** @var array<int, self> */
+    private static array $cleanup = [];
+
+    private ?WeakReference $reference = null;
+    private array $ids = [];
+
+    private function __construct()
+    {
+    }
+
+    public function __destruct()
+    {
+        if (!self::$handlers) {
+            return;
+        }
+        foreach ($this->ids as $id) {
+            unset(self::$handlers[$id]);
+        }
+    }
+
+    public static function register(Closure $closure): void
+    {
+        self::registerShutdownFunction();
+        self::expunge();
+        self::$handlers[] = self::weaken($closure, $target);
+
+        if (!$target) {
+            return;
+        }
+
+        $handlerId = array_key_last(self::$handlers);
+
+        if (PHP_VERSION_ID < 80000) {
+            $cleanup = self::$cleanup[spl_object_id($target)] ??= new self();
+            $cleanup->reference = WeakReference::create($target);
+        } else {
+            /** @phan-suppress-next-line PhanUndeclaredClassReference */
+            assert(class_exists(\WeakMap::class, false));
+            /** @phan-suppress-next-line PhanUndeclaredClassMethod */
+            $weakMap = self::$weakMap ??= new \WeakMap();
+            $object = $target;
+            $cleanup = $weakMap[$object] ??= new self();
+        }
+        $cleanup->ids[] = $handlerId;
+    }
+
+    /**
+     * @see https://github.com/amphp/amp/blob/f682341c856b1f688026f787bef4f77eaa5c7970/src/functions.php#L140-L191
+     */
+    private static function weaken(Closure $closure, ?object &$target = null): Closure
```

Comment:
This is also very useful for the metrics API; we can add a `$weak` parameter to asynchronous instruments `::observe()` method that would allow users to provide normal callbacks with bound `$this` instead of forcing them to take care of weakening the `$this` reference themselves.